File: 061_Reverse Linked List II.py
"""
Reverse Linked List II

Reverse the nodes of a linked list from position left to right (1-indexed).
"""

class ListNode:
    def __init__(self,x):
        self.val = x
        self.next = None

def reverse_between(head: ListNode, left: int, right: int) -> ListNode:
    if not head or left == right: return head
    dummy = ListNode(0); dummy.next = head
    prev = dummy
    for _ in range(left-1): prev = prev.next
    cur = prev.next
    for _ in range(right-left):
        tmp = cur.next
        cur.next = tmp.next
        tmp.next = prev.next
        prev.next = tmp
    return dummy.next

def print_list(h: ListNode):
    out = []
    while h:
        out.append(str(h.val)); h = h.next
    print(' '.join(out))

if __name__ == '__main__':
    a = ListNode(1); a.next = ListNode(2); a.next.next = ListNode(3); a.next.next.next = ListNode(4); a.next.next.next.next = ListNode(5)
    print('RevII 1:', end=' '); print_list(reverse_between(a,2,4))
    b = ListNode(5)
    print('RevII 2:', end=' '); print_list(reverse_between(b,1,1))

# Time Complexity: O(n)
# Space Complexity: O(1)

File: 061_Reverse Linked List II.cpp
/*
Reverse Linked List II

Reverse the nodes of a linked list from position left to right (1-indexed).
*/

#include <bits/stdc++.h>
using namespace std;

struct ListNode { int val; ListNode* next; ListNode(int x): val(x), next(nullptr) {} };

ListNode* reverseBetween(ListNode* head, int left, int right) {
    if (!head || left == right) return head;
    ListNode dummy(0); dummy.next = head;
    ListNode* prev = &dummy;
    for (int i = 1; i < left; ++i) prev = prev->next;
    ListNode* cur = prev->next;
    for (int i = 0; i < right - left; ++i) {
        ListNode* tmp = cur->next;
        cur->next = tmp->next;
        tmp->next = prev->next;
        prev->next = tmp;
    }
    return dummy.next;
}

void printList(ListNode* h) { while (h) { cout << h->val; if (h->next) cout << ' '; h = h->next; } cout << endl; }

int main() {
    ListNode* a = new ListNode(1);
    a->next = new ListNode(2); a->next->next = new ListNode(3);
    a->next->next->next = new ListNode(4); a->next->next->next->next = new ListNode(5);
    auto r = reverseBetween(a, 2, 4); // expected 1 4 3 2 5
    cout << "RevII 1: "; printList(r);

    ListNode* b = new ListNode(5);
    cout << "RevII 2: "; printList(reverseBetween(b,1,1));
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)
